File: src/test/java/com/mpe85/grampa/GrampaTest.java
package com.mpe85.grampa;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import com.mpe85.grampa.exception.ParserCreateException;
import com.mpe85.grampa.rule.Rule;
import com.mpe85.grampa.rule.impl.CharPredicateRule;
import com.mpe85.grampa.rule.impl.EmptyRule;
import com.mpe85.grampa.rule.impl.FirstOfRule;
import com.mpe85.grampa.rule.impl.SequenceRule;

public class GrampaTest {
	
	@Test
	public void createParser_valid_noArgs() {
		final TestParser p = Grampa.createParser(TestParser.class);
		assertNull(p.getDummy());
		
		final Rule<String> root = p.root();
		verifyTestParserRules(root);
	}
	
	@Test
	public void createParser_invalid_privateRuleMethod() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(InvalidParser.class));
	}
	
	@Test
	public void createParser_valid_withArgs() {
		final TestParser p = Grampa.createParser(TestParser.class, String.class).withArgs("foo");
		assertEquals("foo", p.getDummy());
		
		final Rule<String> root = p.root();
		verifyTestParserRules(root);
	}
	
	@Test
	public void createParser_invalid_ctor() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(TestParser.class, Integer.class).withArgs(4711));
	}
	
	@Test
	public void createParser_invalid_args() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(TestParser.class, String.class).withArgs(4711));
	}
	
	@Test
	public void createParser_invalid_noDefaultCtor() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(NoDefaultCtorTestParser.class));
	}
	
	@Test
	public void createParser_invalid_finalRuleMethod() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(FinalRuleMethodTestParser.class));
	}
	
	@Test
	public void createParser_invalid_staticRuleMethod() {
		assertThrows(
				ParserCreateException.class,
				() -> Grampa.createParser(StaticRuleMethodTestParser.class));
	}
	
	@Test
	public void createParser_valid_inheritance() {
		final SuperParser superParser = Grampa.createParser(SuperParser.class);
		verifySuperParserRules(superParser.root());
		
		final SubParser subParser = Grampa.createParser(SubParser.class);
		verifySuperParserRules(subParser.root());
	}
	
	private void verifyTestParserRules(final Rule<String> root) {
		assertTrue(root instanceof FirstOfRule);
		assertEquals(2, root.getChildren().size());
		assertTrue(root.getChildren().get(0) instanceof CharPredicateRule);
		assertTrue(root.getChildren().get(1) instanceof SequenceRule);
		
		final SequenceRule<String> sequenceRule = (SequenceRule<String>) root.getChildren().get(1);
		assertEquals(6, sequenceRule.getChildren().size());
		assertTrue(sequenceRule.getChildren().get(0) instanceof EmptyRule);
		assertTrue(sequenceRule.getChildren().get(1) instanceof EmptyRule);
		assertTrue(sequenceRule.getChildren().get(2) instanceof EmptyRule);
		assertTrue(sequenceRule.getChildren().get(3) instanceof CharPredicateRule);
		assertTrue(sequenceRule.getChildren().get(4) instanceof FirstOfRule);
		assertTrue(sequenceRule.getChildren().get(5) instanceof CharPredicateRule);
		
		assertEquals(root, sequenceRule.getChildren().get(4));
	}
	
	private void verifySuperParserRules(final Rule<String> root) {
		assertTrue(root instanceof SequenceRule);
		assertEquals(2, root.getChildren().size());
		assertTrue(root.getChildren().get(0) instanceof CharPredicateRule);
		assertTrue(root.getChildren().get(1) instanceof SequenceRule);
		
		assertEquals(root, root.getChildren().get(1));
	}
}

File: src/test/kotlin/com/mpe85/grampa/GrampaTest.kt
package com.mpe85.grampa

import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.rule.impl.CharPredicateRule
import com.mpe85.grampa.rule.impl.EmptyRule
import com.mpe85.grampa.rule.impl.FirstOfRule
import com.mpe85.grampa.rule.impl.SequenceRule
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class GrampaTest {

  @Test
  fun createParser_valid_noArgs() {
    val p = TestGrammar::class.createGrammar()
    assertNull(p.dummy)
    val root = p.root()
    verifyTestParserRules(root)
  }

  @Test
  fun createParser_invalid_privateRuleMethod() {
    assertThrows(IllegalArgumentException::class.java) { InvalidGrammar::class.createGrammar() }
  }

  @Test
  fun createParser_valid_withArgs() {
    val p = TestGrammar::class.createGrammar("foo")
    assertEquals("foo", p.dummy)
    val root = p.root()
    verifyTestParserRules(root)
  }

  @Test
  fun createParser_invalid_ctor() {
    assertThrows(IllegalArgumentException::class.java) {
      TestGrammar::class.createGrammar(4711)
    }
  }

  @Test
  fun createParser_invalid_args() {
    assertThrows(IllegalArgumentException::class.java) {
      TestGrammar::class.createGrammar(4711)
    }
  }

  @Test
  fun createParser_invalid_noDefaultCtor() {
    assertThrows(IllegalArgumentException::class.java) { NoDefaultCtorTestGrammar::class.createGrammar() }
  }

  @Test
  fun createParser_invalid_finalRuleMethod() {
    assertThrows(IllegalArgumentException::class.java) { FinalRuleMethodTestGrammar::class.createGrammar() }
  }

  @Test
  fun createParser_invalid_staticRuleMethod() {
    assertThrows(IllegalArgumentException::class.java) { StaticRuleMethodTestGrammar::class.createGrammar() }
  }

  @Test
  fun createParser_valid_inheritance() {
    val superParser = SuperGrammar::class.createGrammar()
    verifySuperParserRules(superParser.root())
    val subParser = SubGrammar::class.createGrammar()
    verifySuperParserRules(subParser.root())
  }

  private fun verifyTestParserRules(root: Rule<String>) {
    assertTrue(root is FirstOfRule<*>)
    assertEquals(2, root.children.size)
    assertTrue(root.children[0] is CharPredicateRule<*>)
    assertTrue(root.children[1] is SequenceRule<*>)
    val sequenceRule = root.children[1] as SequenceRule<String>
    assertEquals(6, sequenceRule.children.size)
    assertTrue(sequenceRule.children[0] is EmptyRule<*>)
    assertTrue(sequenceRule.children[1] is EmptyRule<*>)
    assertTrue(sequenceRule.children[2] is EmptyRule<*>)
    assertTrue(sequenceRule.children[3] is CharPredicateRule<*>)
    assertTrue(sequenceRule.children[4] is FirstOfRule<*>)
    assertTrue(sequenceRule.children[5] is CharPredicateRule<*>)
    assertEquals(root, sequenceRule.children[4])
  }

  @Test
  fun createParser_invalid_vararg() {
    assertThrows(IllegalArgumentException::class.java) { VarArgsTestGrammar::class.createGrammar() }
  }

  private fun verifySuperParserRules(root: Rule<String>) {
    assertTrue(root is SequenceRule<*>)
    assertEquals(2, root.children.size)
    assertTrue(root.children[0] is CharPredicateRule<*>)
    assertTrue(root.children[1] is SequenceRule<*>)
    assertEquals(root, root.children[1])
  }
}
